packages/rpc-methods/src/restricted/notify.ts

```diff
@@ -0,0 +1,133 @@
+   */
+  type: NotificationType;
+
+  /**
+   * A message to show on the notification.
+   */
+  message: string;
+};
+
+export type NotifyMethodHooks = {
+  /**
+   * @param snapId - The ID of the Snap that created the notification.
+   * @param args - The notification arguments.
+   */
+  showNotification: (
+    snapId: string,
+    args: NotificationArgs,
+  ) => Promise<{ isRateLimited: boolean }>;
+};
+
+type SpecificationBuilderOptions = {
+  allowedCaveats?: Readonly<NonEmptyArray<string>> | null;
+  methodHooks: NotifyMethodHooks;
+};
+
+type Specification = ValidPermissionSpecification<{
+  permissionType: PermissionType.RestrictedMethod;
+  targetKey: typeof methodName;
+  methodImplementation: ReturnType<typeof getImplementation>;
+  allowedCaveats: Readonly<NonEmptyArray<string>> | null;
+}>;
+
+const specificationBuilder: PermissionSpecificationBuilder<
+  PermissionType.RestrictedMethod,
+  SpecificationBuilderOptions,
+  Specification
+> = ({ allowedCaveats = null, methodHooks }: SpecificationBuilderOptions) => {
+  return {
+    permissionType: PermissionType.RestrictedMethod,
+    targetKey: methodName,
+    allowedCaveats,
+    methodImplementation: getImplementation(methodHooks),
+  };
+};
+
+export const notifyBuilder = Object.freeze({
+  targetKey: methodName,
+  specificationBuilder,
+  methodHooks: {
+    showNotification: true,
+  },
+} as const);
+
+function getImplementation({ showNotification }: NotifyMethodHooks) {
+  return async function implementation(
+    args: RestrictedMethodOptions<[NotificationArgs]>,
+  ): Promise<null> {
+    const {
+      params,
+      context: { origin },
+    } = args;
+
+    const validatedParams = getValidatedParams(params);
+    const result = await showNotification(origin, validatedParams);
+
+    if (result.isRateLimited) {
+      throw ethErrors.rpc.limitExceeded({
+        message: `Notification type: ${validatedParams.type} is currently rate-limited. Please try again later.`,
+      });
+    }
+
+    return null;
+  };
+}
+
+/**
+ * Validates the notify method `params` and returns them cast to the correct
+ * type. Throws if validation fails.
+ *
+ * @param params - The unvalidated params object from the method request.
+ * @returns The validated method parameter object.
+ */
+function getValidatedParams(params: unknown): NotificationArgs {
+  if (!Array.isArray(params) || !isPlainObject(params[0])) {
+    throw ethErrors.rpc.invalidParams({
+      message: 'Expected array params with single object.',
+    });
+  }
+
+  const { type, message } = params[0];
+
+  if (
+    !type ||
+    typeof type !== 'string' ||
+    !Object.values(NotificationType)
+      .map((n) => n.toString())
+      .includes(type)
+  ) {
+    throw ethErrors.rpc.invalidParams({
+      message: 'Must specify a valid notification "type".',
+    });
+  }
+
+  // @todo Choose sane message limit
+  if (!message || typeof message !== 'string' || message.length >= 200) {
```

Comment:
Suggestion: 140 characters, one old-school tweet's length? :smile:

Reply:
Wait! Could we make this configurable via the specification builder?

Reply:
As agreed on Slack this is not currently attainable.